[EPMEDU-881]: I will feed screen as Dialog

carlossantiagomenendez opened the pull request:
According to https://jira.epam.com/jira/browse/EPMEDU-881

- Create FeedingConfirmationScreen
- "Will Feed" flow suggestion (considering this screen will be called from Home, Search and Favourites)

Video:
https://user-images.githubusercontent.com/110670327/197228244-af758afe-eced-4fdf-a067-a32b047bb320.mp4

It is just an approximation, let me know your ideas about where to put this classes and structure this 'will feed' flow

carlossantiagomenendez committed:
EPMEDU-881: Create ConfirmFeedingScreen

carlossantiagomenendez committed:
EPMEDU-881: Develop changes

carlossantiagomenendez committed:
EPMEDU-881: WillFeedConfirmation navigation to be discussed

carlossantiagomenendez committed:
EPMEDU-881: detekt fixes

Fileek commented on shared/feature/willfeed/src/main/java/com/epmedu/animeal/willfeed/presentation/FeedConfirmationUI.kt:
```diff
@@ -0,0 +1,84 @@
+        Column(
+            modifier = Modifier
+                .fillMaxSize()
+                .padding(24.dp),
```
Can you please align paddings with designs and add status bar padding?

Fileek replied:
Please leave 24 dp for horizontal padding and set vertical one to 40

Fileek commented on shared/feature/willfeed/src/main/java/com/epmedu/animeal/willfeed/presentation/FeedConfirmationUI.kt:
```diff
@@ -0,0 +1,84 @@
+
+@AnimealPreview
+@Composable
+fun Preview() {
```
All previews must be private

Fileek commented on shared/feature/willfeed/src/main/java/com/epmedu/animeal/willfeed/presentation/FeedConfirmationUI.kt:
```diff
@@ -0,0 +1,84 @@
+@Composable
+fun Preview() {
+    AnimealTheme {
+        FeedConfirmationUI(onAgreeClick = { /*TODO*/ }) {
```
Todo is redundant here

Fileek commented on shared/feature/willfeed/src/main/java/com/epmedu/animeal/willfeed/presentation/FeedConfirmationUI.kt:
```diff
@@ -0,0 +1,84 @@
+                AnimealSecondaryButton(
+                    modifier = Modifier.weight(1F),
+                    text = stringResource(id = com.epmedu.animeal.resources.R.string.cancel),
+                    onClick = { onCancelClick.invoke() }
```
You can pass lambdas to onClick as they are if they have no parameters and return unit. Like this: onClick = onCancelClick

carlossantiagomenendez replied:
indeed

Fileek commented on shared/feature/willfeed/src/main/java/com/epmedu/animeal/willfeed/presentation/FeedConfirmationUI.kt:
```diff
@@ -0,0 +1,84 @@
+                AnimealButton(
+                    modifier = Modifier.weight(1F),
+                    text = stringResource(id = com.epmedu.animeal.resources.R.string.agree),
+                    onClick = { onAgreeClick.invoke() }
```
Same as above

Fileek commented on shared/feature/willfeed/src/main/java/com/epmedu/animeal/willfeed/presentation/FeedConfirmationUI.kt:
```diff
@@ -0,0 +1,84 @@
+                )
+                AnimealButton(
+                    modifier = Modifier.weight(1F),
+                    text = stringResource(id = com.epmedu.animeal.resources.R.string.agree),
```
Same as above

Fileek commented on shared/feature/willfeed/src/main/java/com/epmedu/animeal/willfeed/presentation/FeedConfirmationUI.kt:
```diff
@@ -0,0 +1,84 @@
+            ) {
+                AnimealSecondaryButton(
+                    modifier = Modifier.weight(1F),
+                    text = stringResource(id = com.epmedu.animeal.resources.R.string.cancel),
```
Same as above

Fileek commented on shared/feature/willfeed/src/main/java/com/epmedu/animeal/willfeed/presentation/FeedConfirmationUI.kt:
```diff
@@ -0,0 +1,84 @@
+            )
+            Text(
+                modifier = Modifier.padding(top = 36.dp),
+                text = stringResource(id = com.epmedu.animeal.resources.R.string.willfeed_timeleft_msg),
```
Please import R class and you don't need to use named argument for id

carlossantiagomenendez replied:
I take the strings from the resources module and the drawables and mipmaps from the local module. Should I move all the images to the resources module?

Fileek replied:
Yes, we agreed that we will have all resources in the one place

Fileek commented on shared/feature/willfeed/src/main/java/com/epmedu/animeal/willfeed/presentation/FeedConfirmationUI.kt:
```diff
@@ -0,0 +1,84 @@
+import com.epmedu.animeal.willfeed.R
+
+@Composable
+fun FeedConfirmationUI(
```
Please don't forget to add internal modifier if you don't want to share composable to other modules

Fileek commented on shared/feature/willfeed/build.gradle.kts:
```diff
@@ -0,0 +1,29 @@
+}
+
+android {
+    namespace = "com.epmedu.animeal.willfeed"
```
Can you rename the package and module to feedconfirmation to follow consistency?

carlossantiagomenendez replied:
no problem

Fileek commented on shared/feature/willfeed/build.gradle.kts:
```diff
@@ -0,0 +1,29 @@
+    debugImplementation(libs.compose.ui.tooling)
+
+    implementation(libs.hilt.android)
+    kapt(libs.hilt.compiler)
```
Please use only needed dependencies, you don't need all of these. And don't forget to check plugins

Fileek replied:
kapt, datastore, viewmodel and material dialog are also redundant dependencies

Fileek commented on library/navigation/src/main/java/com/epmedu/animeal/navigation/route/WillFeedRoute.kt:
```diff
@@ -0,0 +1,6 @@
+package com.epmedu.animeal.navigation.route
+
+enum class WillFeedRoute {
```
Can you extract Route class from NavigationTab, rename it to TabsRoute and add the FeedConfirmation entry for your screen?

Fileek commented on feature/home/src/main/java/com/epmedu/animeal/home/presentation/HomeScreen.kt:
```diff
@@ -24,7 +28,9 @@ fun HomeScreen() {
         bottomSheetState = bottomSheetState,
         onScreenEvent = viewModel::handleEvents
-    )
+    ) {
```
Please use named argument

carlossantiagomenendez replied:
There is no argument used.

Fileek replied:
onWillFeedClick is a parameter and you pass lambda as a not named argument

carlossantiagomenendez replied:
Do you mean like this?

```
HomeScreenUI(
        state = state,
        bottomSheetState = bottomSheetState,
        onScreenEvent = viewModel::handleEvents,
        onWillFeedClick = {
            navigator.navigate(TabsRoute.FeedConfirmation.name)
        }
    )
```

Fileek replied:
Exactly

Fileek commented on feature/tabsflow/tabs/src/main/java/com/epmedu/animeal/tabs/TabsHost.kt:
```diff
@@ -86,5 +90,24 @@ private fun NavigationTabs(navigationController: NavHostController) {
+                    }
+                )
+            }
+        }
```
I don't see why did you implement all this logic, if cancel button should just use popBackStack and agree button always leads to the same screen

carlossantiagomenendez replied:
You are right. I changed the logic and the parameter from Boolean to Int representing the id of the feeding point selected. This should be useful when the user presses **I will feed button** from search and favourites tab so the HomeScreen knows that it has to change their state to **_showing path to feeding point_** or something like that.

carlossantiagomenendez replied:
Actually, we are going to leave that parameter thing for further implementations

Fileek replied:
We strive to avoid passing parameters in our screen composables. Incapsulate this logic inside the screen implementaion and if we need some branching we will think about some better solution

carlossantiagomenendez replied:
What do you think about this:

```
@Composable
private fun NavigationTabs(navigationController: NavHostController) {
    ScreenNavHost(
        navController = navigationController,
        startDestination = NavigationTab.Home.route.name
    ) {
        screen(TabsRoute.Search.name) { SearchScreen() }
        screen(TabsRoute.Favourites.name) { FavouritesScreen() }
        screen(TabsRoute.Home.name) { HomeScreen() }
        screen(TabsRoute.Analytics.name) { AnalyticsScreen() }
        screen(TabsRoute.More.name) { MoreHost() }
        screen(TabsRoute.FeedConfirmation.name) {
            FeedConfirmationScreen(
                onAgreeClick = { navigationController.navigate(NavigationTab.Home.route.name) },
                onCancelClick = { navigationController.popBackStack() }
            )
        }
    }
}
```

Fileek replied:
Routes are good, but parameters you should remove from FeedConfirmationScreen and use LocalNavigator inside it to navigate

Fileek reviewed (changes requested):
Please put all resources in the resources module

carlossantiagomenendez committed:
EPMEDU-881: code review changes

carlossantiagomenendez committed:
EPMEDU-881: detekt issues

Fileek commented on feature/tabsflow/tabs/src/main/java/com/epmedu/animeal/tabs/TabsHost.kt:
```diff
@@ -86,5 +88,11 @@ private fun NavigationTabs(navigationController: NavHostController) {
         screen(NavigationTab.Home.route.name) { HomeScreen() }
         screen(NavigationTab.Analytics.route.name) { AnalyticsScreen() }
         screen(NavigationTab.More.route.name) { MoreHost() }
```
Let's  change route retrieving to TabsRoute instead of NavigationTab

Fileek commented on library/navigation/src/main/java/com/epmedu/animeal/navigation/route/TabsRoute.kt:
```diff
@@ -0,0 +1,10 @@
+package com.epmedu.animeal.navigation.route
```
Good, but we forgot to move SignUpRoute to common module. Can you move TabsRoute over there please?

Fileek commented on shared/feature/feedconfirmation/src/main/java/com/epmedu/animeal/feedconfirmation/presentation/FeedConfirmationUI.kt:
```diff
@@ -0,0 +1,87 @@
+) {
+    Scaffold(
+        modifier = Modifier
+            .statusBarsPadding()
```
You must use status bar padding on the top element of the screen, not on the root element, otherwise there will be status bar blinking on navigation between screens

carlossantiagomenendez replied:
Copy that. Should I put it on the Column element then?

Fileek replied:
Yep

Fileek commented on shared/feature/feedconfirmation/src/main/java/com/epmedu/animeal/feedconfirmation/presentation/FeedConfirmationUI.kt:
```diff
@@ -0,0 +1,87 @@
+                modifier = Modifier.fillMaxWidth(),
+                horizontalArrangement = Arrangement.spacedBy(16.dp)
+            ) {
+                AnimealSecondaryButton(
```
Please change it to outlined version according to designs

carlossantiagomenendez committed:
EPMEDU-881: code review improvements

carlossantiagomenendez commented:
Ready to review!

carlossantiagomenendez committed:
EPMEDU-881: adjust some paddings

carlossantiagomenendez committed:
EPMEDU-881: Transform screen to dialog

carlossantiagomenendez commented:
Changes done:

- Added FullScreenDialog to foundation module.
- Refactor FeedConfirmationScreen to dialog.

At the moment it does not affect other dialogs(log out dialog) in the app despite changing themes configuration.

https://user-images.githubusercontent.com/110670327/198273084-fb72a178-0cf0-4ec4-b2e6-9730b0aaeed0.mp4

Ready to review!

Fileek commented on feature/home/src/main/java/com/epmedu/animeal/home/presentation/HomeScreen.kt:
```diff
@@ -1,35 +1,30 @@
-    val navigator = LocalNavigator.currentOrThrow
 
     val state by viewModel.stateFlow.collectAsState()
+    val feedDialogState: MutableState<Boolean> = remember { mutableStateOf(false) }
```
You can change it to var and use by keyword instead of = to avoid writing .value every time

Fileek commented on library/common/src/main/java/com/epmedu/animeal/common/route/TabsRoute.kt:
```diff
@@ -0,0 +1,10 @@
+    Home,
+    Analytics,
+    More,
+    FeedConfirmation
```
Now it's redundant)

carlossantiagomenendez committed:
EPMEDU-881: Code review improvements, screen as dialog behaviour

carlossantiagomenendez commented:
Ready to review!

Fileek commented on shared/feature/feedconfirmation/src/main/java/com/epmedu/animeal/feedconfirmation/presentation/FeedConfirmationDialog.kt:
```diff
@@ -0,0 +1,94 @@
+    onAgreeClick: () -> Unit,
+    onCancelClick: () -> Unit
+) {
+    FeedConfirmationUI(onAgreeClick, onCancelClick)
```
Let's not wrap UI content in a separate composable, because it's not a screen, but a UI component.

carlossantiagomenendez replied:
How should I name the final composable function: FeedConfirmationUI or FeedConfirmationDialog?

Fileek replied:
> How should I name the final composable function: FeedConfirmationUI or FeedConfirmationDialog?

FeedConfirmationDialog

carlossantiagomenendez committed:
EPMEDU-881: Rename FeedConfirmationUI.kt to FeedConfirmationDialog.kt

VEZE commented on feature/home/src/main/java/com/epmedu/animeal/home/presentation/HomeScreen.kt:
```diff
@@ -1,11 +1,9 @@
-import androidx.compose.runtime.LaunchedEffect
-import androidx.compose.runtime.collectAsState
-import androidx.compose.runtime.getValue
+import androidx.compose.runtime.*
```
Please setup kotlin codestyle to not use wildcards

VEZE replied:
Remove wildcards

carlossantiagomenendez replied:
Is it possible that some CI job did this? Because in my repository is not showing like that

VEZE replied:
Can you check commit with changing those rows in detekt config?
NoWildcardImports:
    active: true

VEZE commented on feature/home/src/main/java/com/epmedu/animeal/home/presentation/HomeScreenUI.kt:
```diff
@@ -29,6 +29,7 @@ internal fun HomeScreenUI(
     state: HomeState,
     bottomSheetState: HomeBottomSheetState,
     onScreenEvent: (HomeScreenEvent) -> Unit,
+    onWillFeedClick: () -> Unit
```
May be we should add it to onScreenEvent?

carlossantiagomenendez replied:
That is a good suggestion but the Dialog must be used from Search and Favourites also...

Fileek replied:
It is a good suggestion, you just need to create WillFeedClick event for the screen, ShowConfirmationDialog event for viewmodel and handle both. No conflicts with other flows, because we don't change the dialog logic, only the screen logic.

VEZE reviewed (changes requested):
Why png resource?

carlossantiagomenendez commented:
> Why png resource?

What do you suggest to use instead, a vector image? I think I've tried it that way but android studio did not processed correctly

VEZE commented:
> > Why png resource?
> 
> What do you suggest to use instead, a vector image? I think I've tried it that way but android studio did not processed correctly

Vectors is used across all the app, if smth wrong with them, just mention that in figma and w8 for changes.

Fileek commented:
@carlossantiagomenendez I've already reached out to our designer, she'll look into it, and I'll let you know when the image is fixed

Fileek commented:
@carlossantiagomenendez the image is ready, check figma :)

carlossantiagomenendez commented:
> @carlossantiagomenendez the image is ready, check figma :)

Thanks!

carlossantiagomenendez committed:
EPMEDU-881: Change icon resource and move logic to viewmodel

carlossantiagomenendez committed:
EPMEDU-881: Merge with develop

carlossantiagomenendez commented:
Things done:

- Updated the vector resource
- Implemented suggested logic in view model
- Added back handler implementation for dialog 

Let me know if anything is missing :)

VEZE commented on feature/home/src/main/java/com/epmedu/animeal/home/presentation/viewmodel/HomeViewModel.kt:
```diff
@@ -40,6 +40,8 @@ class HomeViewModel @Inject constructor(
         is HomeScreenEvent.FeedingPointSelected -> selectFeedingPoint(event)
         is HomeScreenEvent.FeedingPointFavouriteChange -> changeFavouriteFeedingPoint(event)
         is HomeScreenEvent.UserCurrentGeolocationRequest -> changeGpsSetting()
+        is HomeScreenEvent.ShowWillFeedDialog -> updateWillFeedDialog(true)
```
WillFeedDialog.show() or showWillFeedDialog separate methods

carlossantiagomenendez replied:
Ok, I did it cause eventually detekt will complain about max number of methods(14)

VEZE commented on feature/home/src/main/java/com/epmedu/animeal/home/presentation/viewmodel/HomeState.kt:
```diff
@@ -13,5 +13,6 @@ data class HomeState(
     val gpsSettingState: GpsSettingState = GpsSettingState.Disabled,
-    val feedingPoints: StableList<FeedingPointUi> = StableList(emptyList())
+    val feedingPoints: StableList<FeedingPointUi> = StableList(emptyList()),
+    val isWillFeedDialogShowing: Boolean = false
```
Change to State

carlossantiagomenendez committed:
EPMEDU-881: Code review

carlossantiagomenendez commented:
Ready to review!

Fileek committed:
Merge branch 'develop' into feature/EPMEDU-881_I_will_feed
